url49584c1d41_Arrem_GGS_tgz-pJ8-com_gamezgalaxy_test_console_MainJ8/src/com/gamezgalaxy/GGS/networking/packets/minecraft/GlobalPosUpdate.java

```diff
@@ -0,0 +1,138 @@
+/*******************************************************************************
+ * Copyright (c) 2012 GamezGalaxy.
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the GNU Public License v3.0
+ * which accompanies this distribution, and is available at
+ * http://www.gnu.org/licenses/gpl.html
+ ******************************************************************************/
+package com.gamezgalaxy.GGS.networking.packets.minecraft;
+
+import java.io.IOException;
+
+import com.gamezgalaxy.GGS.API.io.PacketPrepareEvent;
+import com.gamezgalaxy.GGS.iomodel.Player;
+import com.gamezgalaxy.GGS.networking.IOClient;
+import com.gamezgalaxy.GGS.networking.packets.Packet;
+import com.gamezgalaxy.GGS.networking.packets.PacketManager;
+import com.gamezgalaxy.GGS.networking.packets.PacketType;
+import com.gamezgalaxy.GGS.server.Server;
+
+public class GlobalPosUpdate extends Packet {
+
+	public GlobalPosUpdate(String name, byte ID, PacketManager parent,
+			PacketType packetType) {
+		super(name, ID, parent, packetType);
+		// TODO Auto-generated constructor stub
+	}
+	
+	public GlobalPosUpdate(PacketManager pm) {
+		super("GlobalPosUpdate", (byte)0x09, pm, PacketType.Server_to_Client);
+	}
+	
+	public byte[] toSend(Player p) {
+		byte[] finals;
+		if (posUpdate(p) && rotUpdate(p)) {
+			finals = new byte[7];
+			finals[0] = ID;
+			finals[1] = p.getID();
+			finals[2] = (byte)(p.getX() - p.oldX);
+			finals[3] = (byte)(p.getY() - p.oldY);
+			finals[4] = (byte)(p.getZ() - p.oldZ);
+			finals[5] = p.yaw;
+			finals[6] = p.pitch;
+		}
+		else if (posUpdate(p)) {
+			finals = new byte[5];
+			finals[0] = (byte)0x0a;
+			finals[1] = p.getID();
+			finals[2] = (byte)(p.getX() - p.oldX);
+			finals[3] = (byte)(p.getY() - p.oldY);
+			finals[4] = (byte)(p.getZ() - p.oldZ);
+		}
+		else if (rotUpdate(p)) {
+			finals = new byte[4];
+			finals[0] = (byte)0x0b;
+			finals[1] = p.getID();
+			finals[2] = p.yaw;
+			finals[3] = p.pitch;
+		}
+		else { //PING!
+			finals = new byte[1];
+			finals[0] = (byte)0x01;
+		}
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Similar code fragments were detected in the same file at the following lines: 44:62, 94:112.  
Refactoring can help improve code maintainability. Consider reducing duplicate code by extracting it into a separate method. You can then replace duplicated code with calls to this new method.